File: catkin_ws/src/wave_judger/src/image_subscriber.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include "wave_judger.hpp"
#include "OpenPose/Rect.h"
#include "OpenPose/RectArray.h"
#include "std_msgs/String.h"

WaveJudger wj;
ros::Publisher *msg_pub_;
cv::Mat *dep_img = NULL;

int succ = -10;
bool swit = 0;

void ImageCallback(const sensor_msgs::ImageConstPtr &msg){
	// if(!dep_img || succ == 0)return;
    // if(succ == 0)return;
    try{
        // cv::Mat *depth = new cv::Mat;
        // *depth = cv_bridge::toCvShare(msg, "bgr8")->image;//sensor_msgs::image_encodings::TYPE_32FC1)->image;
        // cv::imshow("", *depth);
        // cv::waitKey(0);
        // printf("get msgs.\n");
        auto img = cv_bridge::toCvShare(msg, "bgr8")->image;
        float box[20][4];
        int n_box = wj.FindWaving(img, box, swit);
        OpenPose::RectArray bboxes;
        OpenPose::Rect rect;
        // printf("Get");
        if(n_box > 0){
            // printf("Sending");
            for(int i = 0; i < n_box; i++){
                rect.x0 = box[i][0];
                rect.y0 = box[i][1];
                rect.x1 = box[i][2];
                rect.y1 = box[i][3];
                bboxes.bboxes.push_back(rect);
            }
            // for(int j = 0; j < 10; j++){
                msg_pub_->publish(bboxes);
            // }
            succ++;// = true;
        }
        wj.vis(cv_bridge::toCvShare(msg, "bgr8")->image, swit);
    }catch(cv_bridge::Exception &e){
        ROS_ERROR("Could not convert from '%s' to 'bgr8'.", msg->encoding.c_str());
    }
}

// void DImageCallback(const sensor_msgs::ImageConstPtr &msg){
//     if(!dep_img)dep_img = new cv::Mat;
//     dep_img = cv_bridge::toCvShare(msg, "mono8")->image;
// }

// void CtrlCallback(const std_msgs::String &msg){
//     if(!swit && !strcmp(msg.data.c_str(), "go back to the serve place")){
//         swit = 1;
//         succ = -10;
//     }
// }
// robot start get coat
int main(int argc, char **argv){
    ros::init(argc, argv, "Image_subscriber");
    ros::NodeHandle nh;

    // auto swit_scb = nh.subscribe("control_command", 1, CtrlCallback);
    
    msg_pub_ = new ros::Publisher;
    *msg_pub_ = nh.advertise<OpenPose::RectArray>("People", 1);
    cv::namedWindow("");
    cv::startWindowThread();

    image_transport::ImageTransport it(nh);
    image_transport::Subscriber sub = it.subscribe("/astra/rgb/image_raw", 1, ImageCallback);

    // image_transport::Subscriber sub_d = it.subscribe("/camera/depth/image_raw", 1, DImageCallback);
    
    auto spinner = ros::MultiThreadedSpinner(3);
    spinner.spin();
    
    cv::destroyAllWindows();

    return 0;
}

File: catkin_ws/src/face_count/scripts/image_subscriber.py
import cv2
import rospy
from sensor_msgs.msg import Image
import numpy as numpy
from cv_bridge import CvBridge, CvBridgeError
import sys

from detect import detectFace

from sound_play.msg import SoundRequest
from sound_play.libsoundplay import SoundClient

def ImageCallback(msg):
    global bridge
    global sc
    global cnt_call
    try:
        img = bridge.imgmsg_to_cv2(msg, 'bgr8')
        boxes = detectFace(img, threshold=[0.6, 0.6, 0.7])
        cnt = len(boxes)
        
        for box in boxes:
            x0, y0, x1, y1 = box[:4]
            cv2.rectangle(img, (int(x0), int(y0)), (int(x1), int(y1)), (255, 0, 0), 3)
        rospy.loginfo('detected %d people.'%(cnt))
        
        if cnt_call % 20 == 0:
            sc.say('detected %d people'%(cnt))

        cv2.imshow('', img)
        cv2.waitKey(1)
    except:
        print("Could not convert from '%s' to 'bgr8'." % str(msg.encoding))
    
    cnt_call += 1

def SubscribImages():
    global cnt_call
    cnt_call = 0
    global bridge
    bridge = CvBridge()

    rospy.init_node('image_subscriber', anonymous=True)
    global sc
    sc = SoundClient()
    rospy.Subscriber('camera/image', Image, ImageCallback)
    rospy.spin()

if __name__ == "__main__":
    SubscribImages()
